### ISSUE TITLE
Desktop: Sync scrolling doesn't map source lines to html blocks

### ISSUE THREAD

**Issue by CalebJohn** [laurent22/joplin#5811]
The sync scrolling does attempt to add source lines to html blocks using the `token.attrJoin` and `token.attrSet` functions. But these attributes are lost because the markdownIt renderer ultimately passes through the html content and doesn't propagate these attributes.

For typical notes, this isn't much of an issue, but for notes that extensively use html, the sync scrolling won't work very well. Fortunately, it won't fall flat on it's face, it will still scroll the same as it did before the sync scrolling update.

In order to improve the handling of these HTML blocks, a new function should be added to `renderer/htmlUtils.ts` that takes in an html string and attaches attributes. This function can then be called in `renderer/MdToHtml/rules/source_map.ts` when dealing with html blocks.

**Comment by laurent22** [comment]
What kind of attributes should be added? Is that the line number or something like this?

**Comment by CalebJohn** [comment]
> What kind of attributes should be added? Is that the line number or something like this?

Yes, it's an attribute called `source-line` and we also add a new class `maps-to-line`.

**Comment by CalebJohn** [comment]
Not stale.

**Comment by laurent22** [comment]
Wasn't this resolve with Kenichi's sync scroll feature?

**Comment by CalebJohn** [comment]
This is a limitation of the current sync scrolling implementation. It relies on attaching some data to the rendered HTML, but when the md file contains html, those blocks are added to the rendered view without the extra data added. The result is that the sync-scroll algorithm must fall back on linear scrolling. I've attached a note that demonstrates this behaviour.
[html_sync_scroll_test.md](https://github.com/laurent22/joplin/files/11750668/html_sync_scroll_test.md)

What this case shows is that there is no syncing happening between the 2 views.
![image](https://github.com/laurent22/joplin/assets/2179547/619e20ba-8450-4360-998f-b112c97b636e)

**Comment by laurent22** [comment]
Right, that makes sense, so indeed it's still an issue. Probably some extra tags need to be injected when the document is being parsed
